pvlib/iotools/psm4.py

```diff
@@ -740,34 +731,42 @@ def parse_nsrdb_psm4(fbuf, map_variables=True):
     .. [2] `Standard Time Series Data File Format
        <https://web.archive.org/web/20170207203107/https://sam.nrel.gov/sites/default/files/content/documents/pdf/wfcsv.pdf>`_
     """
-    # The first 2 lines of the response are headers with metadata
-    metadata_fields = fbuf.readline().split(',')
-    metadata_fields[-1] = metadata_fields[-1].strip()  # strip trailing newline
-    metadata_values = fbuf.readline().split(',')
-    metadata_values[-1] = metadata_values[-1].strip()  # strip trailing newline
-    metadata = dict(zip(metadata_fields, metadata_values))
-    # the response is all strings, so set some metadata types to numbers
-    metadata['Local Time Zone'] = int(metadata['Local Time Zone'])
-    metadata['Time Zone'] = int(metadata['Time Zone'])
-    metadata['Latitude'] = float(metadata['Latitude'])
-    metadata['Longitude'] = float(metadata['Longitude'])
-    metadata['Elevation'] = int(metadata['Elevation'])
-    # get the column names so we can set the dtypes
-    columns = fbuf.readline().split(',')
-    columns[-1] = columns[-1].strip()  # strip trailing newline
-    # Since the header has so many columns, excel saves blank cols in the
-    # data below the header lines.
-    columns = [col for col in columns if col != '']
-    dtypes = dict.fromkeys(columns, float)  # all floats except datevec
-    dtypes.update(Year=int, Month=int, Day=int, Hour=int, Minute=int)
-    dtypes['Cloud Type'] = int
-    dtypes['Fill Flag'] = int
-    data = pd.read_csv(
-        fbuf, header=None, names=columns, usecols=columns, dtype=dtypes,
-        delimiter=',', lineterminator='\n')  # skip carriage returns \r
+    if hasattr(filename, "read"):
+        # already a file-like object
+        context = contextlib.nullcontext(filename)
+    else:
+        # otherwise, assume a filename or path
+        context = open(str(filename), 'r')
```

Comment:
is this if/else pattern going to be repeated throughout iotools? If so we should have a private function in a central place.